File: examples/04-java-threads/Example8.java
// =================================================================
//
// File: Example8.java
// Author: Sandra Tello A01703658 Isaac Planter A01702962
// Description: This file implements the enumeration sort algorithm.
// 				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2021 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.Arrays;

public class Example8 extends Thread {
	private static final int SIZE = 100_000;
	private int array[], start, end, aux[];
	//private int aux[] = new int[SIZE];

	public Example8(int array[], int start, int end, int aux[]) {
		this.array = array;
		this.start = start;
		this. end = end;
		this.aux=aux;
	}

	public void run(){
		
		for (int i = start; i < end; i++) {
			aux[i] = 0;
		}
		for(int i=start;i<end;i++){
			for(int j=0;j<SIZE;j++){
				if(array[i]>array[j] | (array[i] == array[j] && j < i)){
					aux[i] += 1;
				}
			}
		}
		for (int k = start; k < end; k++) {
			array[aux[k]] = array[k];
		}
	}

	public int[] getSortedArray() {
		return array;
	}




	public static void main(String args[]) {
		int block;
		double ms;
		int array[] = new int[SIZE];
		int aux[]=new int[SIZE];

		long startTime, stopTime;
		Example8 threads[];

		Utils.randomArray(array);
		Utils.displayArray("before", array);
		
		block = SIZE / Utils.MAXTHREADS;
		threads = new Example8[Utils.MAXTHREADS];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;

		for (int j = 1; j <= Utils.N; j++) {
			for (int i = 0; i < threads.length; i++) {
				if (i != threads.length - 1) {
					threads[i] = new Example8(array, (i * block), ((i + 1) * block), aux);
				} else {
					threads[i] = new Example8(array, (i * block), SIZE, aux);
				}
			}

			startTime = System.currentTimeMillis();
			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms +=  (stopTime - startTime);
		}

		Utils.displayArray("after", threads[0].getSortedArray());
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}

File: examples/01-intro-c/example8.c
// =================================================================
//
// File: example8.c
// Author(s): Sandra Tello Salinas e Isaac Planter
// Description: This file contains the code that implements the
//				enumeration sort algorithm. The time this implementation
//				takes ill be used as the basis to calculate the
//				improvement obtained with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"

#define SIZE 10000

// b implement your code
void enumerationSort (int *arr, int size){
	int aux[size];
	for (int i = 0; i < size; i++){
		int menores = 0;
		for (int j = 0; j< size; j++){
			if(arr[i]>arr[j] || (arr[i]==arr[j]&&i<j) ){
				menores+=1;
			}
		}
		aux[menores] = arr[i];
	}
	for (int k = 0; k<size; k++){
		arr[k] = aux[k];
	}
}


int main(int argc, char* argv[]) {
	int i, *a;
	double ms;

	a = (int*) malloc(sizeof(int) * SIZE);
	random_array(a, SIZE);
	display_array("before", a);

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		// call the implemented function
		enumerationSort(a, SIZE);
		ms += stop_timer();
	}
	display_array("after", a);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}
